src/main/scala/ir/cilvisitor/CILVisitor.scala

```diff
@@ -0,0 +1,196 @@
+package ir.cilvisitor
+
+import ir.*
+import scala.collection.mutable.ArrayBuffer
+
+/** A new visitor based off CIL.
+  *
+  * Copied from ASLi https://github.com/UQ-PAC/aslp/blob/partial_eval/libASL/asl_visitor.ml copying from George Necula's
+  * CIL project (https://people.eecs.berkeley.edu/~necula/cil/)
+  */
+
+enum VisitAction[T]:
+  case SkipChildren()
+  case DoChildren()
+  case ChangeTo(e: T)
+  // changes to e, then visits children of e, then applies f to the result
+  case ChangeDoChildrenPost(e: T, f: T => T)
+
+import VisitAction._
+
+trait CILVisitorImpl:
+  def vprog(e: Program): VisitAction[Program] = VisitAction.DoChildren()
+  def vproc(e: Procedure): VisitAction[List[Procedure]] = VisitAction.DoChildren()
+  def vparams(e: ArrayBuffer[Parameter]): VisitAction[ArrayBuffer[Parameter]] = VisitAction.DoChildren()
+  def vblock(e: Block): VisitAction[Block] = VisitAction.DoChildren()
+
+  def vstmt(e: Statement): VisitAction[List[Statement]] = VisitAction.DoChildren()
+  def vjump(j: Jump): VisitAction[Jump] = VisitAction.DoChildren()
+  def vfallthrough(j: Option[GoTo]): VisitAction[Option[GoTo]] = VisitAction.DoChildren()
+
+  def vexpr(e: Expr): VisitAction[Expr] = VisitAction.DoChildren()
+  // right-expressions
+  def vvar(e: Variable): VisitAction[Variable] = VisitAction.DoChildren()
+  def vmem(e: Memory): VisitAction[Memory] = VisitAction.DoChildren()
+  def vload(e: MemoryLoad): VisitAction[MemoryLoad] = VisitAction.DoChildren()
+
+  // left-expressions
+  // Somewhat confusing that these do not get visited by vexpr
+  def vlvar(e: Variable): VisitAction[Variable] = VisitAction.DoChildren()
+  def vlmem(e: Memory): VisitAction[Memory] = VisitAction.DoChildren()
+
+  def enter_scope(params: ArrayBuffer[Parameter]): Unit = ()
+  def leave_scope(outparam: ArrayBuffer[Parameter]) : Unit = ()
+
+
+class CILVisitor(val v: CILVisitorImpl) {
```

Comment:
suggest renaming "CILVisitorImpl", since the CILVisitor seems to have the main implementation, not the other way around. maybe "CILVisitorActions" for Impl?